azure/scope/machinepool.go

```diff
@@ -405,8 +405,27 @@ func (m *MachinePoolScope) applyAzureMachinePoolMachines(ctx context.Context) er
 	}
 
 	existingMachinesByProviderID := make(map[string]infrav1exp.AzureMachinePoolMachine, len(ampms))
-	for _, machine := range ampms {
-		existingMachinesByProviderID[machine.Spec.ProviderID] = machine
+	for _, ampm := range ampms {
+		// propagate Machine delete annotation from owner machine to AzureMachinePoolMachine
+		// this ensures setting a deleteMachine annotation on the Machine has an effect on the AzureMachinePoolMachine
+		// and the deployment strategy.
+		machine, err := util.GetOwnerMachine(ctx, m.client, ampm.ObjectMeta)
+		if err != nil {
+			// TODO(mw): which verbosity? info or error?
+			log.V(4).Info("failed to get owner machine", "machine", ampm.Spec.ProviderID)
+			continue
+		}
+		if machine != nil && machine.Annotations != nil {
+			if _, hasDeleteAnnotation := machine.Annotations[clusterv1.DeleteMachineAnnotation]; hasDeleteAnnotation {
+				log.V(4).Info("propagating DeleteMachineAnnotation", "machine", ampm.Spec.ProviderID)
+				if ampm.Annotations == nil {
+					ampm.Annotations = make(map[string]string)
+				}
+				ampm.Annotations[clusterv1.DeleteMachineAnnotation] = machine.Annotations[clusterv1.DeleteMachineAnnotation]
```

Comment:
I am trying to understand how this annotation is getting consumed. 
Please point me to the right location, @Jont828 and @mweibel. 
I looked at https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/52df930d5c636071091f7ba904b606bc095346c6/azure/scope/machinepool.go#L527, and `DeleteMachine` is attempting to delete ownerMachine of this `AzureMachinePoolMachine`.

I then looked at AzureMachinePool.ReconcileDelete at https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/52df930d5c636071091f7ba904b606bc095346c6/exp/controllers/azuremachinepool_controller.go#L253-L255, and I noticed that an AzureMachinePool resource only gets deleted when its deletionTimestamp is set. 

I am trying to build the knowledge gap where this annotation be used.

Reply:
Adding more context:
`getDeletingMachines` that should delete the underlying ampm if owner machine is deleted, is also looking for a deletion timestamp. Reference: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/52df930d5c636071091f7ba904b606bc095346c6/azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go#L241

Trying to understand this better.

Reply:
My investigation above with Jonathan's comment https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4949#issuecomment-2192548127 is clearing up the picture.


Marking my above comments as resolved.